src/grad.jl

```diff
@@ -1,40 +1,64 @@
 export grad, jacobian, jvp, j′vp, to_vec
+function replace_arg(x, xs::Tuple, k::Int)
+    return ntuple(length(xs)) do p
+        if p == k
+            x
+        else
+            xs[p]
+        end
+    end
+end
 
 """
-    grad(fdm, f, x::Vector{<:Number})
+    grad(fdm, f, xs...)
 
-Approximate the gradient of `f` at `x` using `fdm`. Assumes that `f(x)` is scalar.
+Approximate the gradient of `f` at `xs...` using `fdm`. Assumes that `f(xs...)` is scalar.
 """
-function grad(fdm, f, x::Vector{T}) where T<:Number
-    v, dx, tmp = fill(zero(T), size(x)), similar(x), similar(x)
-    for n in eachindex(x)
-        v[n] = one(T)
-        dx[n] = fdm(function(ϵ)
-                tmp .= x .+ ϵ .* v
-                return f(tmp)
-            end,
-            zero(T),
-        )
-        v[n] = zero(T)
+function grad end
+
+function grad(fdm, f, x::AbstractArray{T}) where T
+    dx, tmp = similar(x), similar(x)
+    for k in eachindex(x)
+        dx[k] = fdm(zero(T)) do ϵ
+            tmp .= x
+            tmp[k] += ϵ
+            return f(tmp)
+        end
     end
     return dx
 end
 
+grad(fdm, f, x::Real) = fdm(f, x)
+grad(fdm, f, x::Tuple) = grad(fdm, (xs...)->f(xs), x...)
+
+function grad(fdm, f, xs...)
+    return ntuple(length(xs)) do k
+        grad(fdm, x->f(replace_arg(x, xs, k)...), xs[k])
+    end
+end
+
 """
-    jacobian(fdm, f, x::Vector{<:Number}, D::Int)
-    jacobian(fdm, f, x::Vector{<:Number})
+    jacobian(fdm, f, xs::Union{Real, AbstractArray{<:Real}}[; dim::Int=length(f(x))])
 
 Approximate the Jacobian of `f` at `x` using `fdm`. `f(x)` must be a length `D` vector. If
 `D` is not provided, then `f(x)` is computed once to determine the output size.
 """
-function jacobian(fdm, f, x::Vector{T}, D::Int) where {T<:Number}
-    J = Matrix{T}(undef, D, length(x))
-    for d in 1:D
-        J[d, :] = grad(fdm, x->f(x)[d], x)
+function jacobian(fdm, f, x::Union{T, AbstractArray{T}}; dim::Int=length(f(x))) where {T <: Real}
+    J = Matrix{float(T)}(undef, dim, length(x))
+    for d in 1:dim
+        gs = grad(fdm, x->f(x)[d], x)
+        for k in 1:length(x)
+            J[d, k] = gs[k]
+        end
     end
     return J
 end
-jacobian(fdm, f, x::Vector{<:Number}) = jacobian(fdm, f, x, length(f(x)))
+
+function jacobian(fdm, f, xs...; dim::Int=length(f(xs...)))
```

Comment:
there is a lot of `...` splatting in the PR. I wonder (haven't checked) if it is worth using `Vararg{<:Any, N} where N`? (for performance https://github.com/JuliaLang/julia/pull/32817/files)

Reply:
I'm not sure. I thought I didn't need to care much about performance in this package (since it was made for test). And I guess it would be fine, since places contains splatting are always tuple and only has a few lines.

Reply:
And I don't think I should use `Vararg{<:Any, N}` this will force all the args to be the same type, which is not correct. It is reasonable to allow functions like

```jl
f(rand(2, 2), 2)
```

to work with `grad` and `jacobian`

Reply:
all good points. I am convinced. thanks! :)